docs/_tutorials/reference.md

```diff
@@ -0,0 +1,138 @@
+| `app.shortcut(callbackId, fn);` | Listens for global or message shortcut invocation. The `callbackId` is a `string` or `RegExp` that must match a shortcut `callback_id` specified within your app's configuration.
+| `app.command(commandName, fn);` | Listens for slash command invocations. The `commandName` is a `string` that must match a slash command specified in your app's configuration. Slash command names should be prefaced with a `/` (ex: `/helpdesk`).
+| `app.view(callbackId, fn);`     | Listens for `view_submission` and `view_closed` events. `view_submission` events are sent when a user submits a modal that your app opened. `view_closed` events are sent when a user closes the modal rather than submits it.
+| `app.options(actionId, fn);`    | Listens for options requests (from select menus with an external data source). This isn't often used, and shouldn't be mistaken with `app.action`. The `actionId` identifier is a `string` that matches the unique `action_id` included when you app sends a [select with an external data source](https://api.slack.com/reference/block-kit/block-elements#external_select).
+
+⚙️ There are also **[globally-scoped middleware functions](/concepts#global-middleware)** that will be run for all incoming events. These can be passed into the `use(fn)` method.
+
+#### Constraint objects
+There are a collection of constraint objects that some methods have access to. These can be used to narrow the event your listener function handles, or to handle special cases. Constraint objects can be passed in lieu of the identifiers outlined above. Below is a collection of constraint objects and the methods they can be passed to. 
+
+| Method                          | Options | Details |
+|--------------------------------------------------------------------|
+| `app.action(constraints, fn)` | `block_id`, `action_id`, `callback_id` | Listens for more than just the `action_id`. `block_id` is the ID for the block that contains the action element. `callback_id` is the `callback_id` of the parent view that is passed when instantiating it. `callback_id`s can only be passed when handling action elements within modals. |
+| `app.shortcut(constraints, fn)` | `type`, `callback_id` | Allows specification of the type of shortcut. `type` must either be `shortcut` for **global shortcuts** or `message_action` for **message_shortcuts**. `callbackId` can be a `string` or `RegExp`. |
+| `app.view(constraints, fn)` | `type`, `callback_id` | `type` must either be `view_closed` or `view_submission`, which determines what specific event your listener function is sent. `callback_id` is the `callback_id` of the view that is sent when your app opens the modal. |
+| `app.action(constraints, fn);` | `type` | Used primarily to listen to legacy interactive message payloads. Type must be `block_actions` or `interactive_message` |
```

Comment:
There are two `app.action` in this table. Can we combine them?

Reply:
I moved them next to each other, but I was hesitant to combine them because they are used for different use cases. Do you know of a way to possibly combine them?

Reply:
>app.action({"type": "block_actions", "action_id": "button"}, async ({ ack, body}) => {
  await ack();
})

The `type` is available only when using along with some ID:
```js
app.action({"type": "block_actions", "action_id": "button"}, async ({ ack, body}) => {
  await ack();
})
```
So, separating `type` from other keys does not make sense to me. That being said, I do understand mixing this with the above one may be confusing (the above one focuses on block_actions). The following idea could require a bit longer description but how about adding `type` in this way?

>| `app.action(constraints, fn)` | `block_id`, `action_id`, `callback_id` (, `type`) | Listens for more than just the `action_id`. `block_id` is the ID for the block that contains the action element. `callback_id` is the `callback_id` of the parent view that is passed when instantiating it. `callback_id`s can only be passed when handling action elements within modals. To specifically handle only either an action in blocks or in attachments, you can use `type` field for it. Possible values for `type` are `block_actions` (for blocks) or `interactive_message` (for attachments) |